### ISSUE TITLE
Support for downloading sources from mirrors

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#328]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/328)
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 29, 2018, 10:18

This describes the client side support for source mirroring as discussed [in this thread](https://mail.gnome.org/archives/buildstream-list/2018-March/msg00004.html). For details on the automation of *project driven mirroring*, see separate issue #330.

This is largely based on the extension of my counter proposal in [this email](https://mail.gnome.org/archives/buildstream-list/2018-March/msg00013.html)

The following draft makes the assumption that a given [source alias](http://buildstream.gitlab.io/buildstream/projectconf.html#source-aliases) should be mirrored as a single unit; which makes the implementation a bit simpler, while imposing that a `project.conf` declare aliases in such a way that matches the chosen mirroring solution. This is to say, if you have a lot of git repositories in the same upstream location, but only wish to mirror a portion of those git repositories, you must use separate *source aliases* to refer to them in your project.

## What is a "mirror"

We are going to treat a "mirror" as an abstract thing, with some elaborate configuration which will allow both interoperability with any mirroring solutions, and also allow for more simplified configuration for a more turn key solution, where BuildStream has an opportunity to provide something more practical and easy to use.

A mirror, in BuildStream terms, is a configuration object which is declared in a `project.conf`, which may define 0 or more mirrors associated to the given project.

Some general rules about a mirror:
* A single mirror is related to a *geographical location*
* The first defined mirror is the default mirror
  * User configuration, or possibly command line options, should allow the user to select a *preferred mirror*
* A mirror defines a single *alias mapping* for a given [source alias](http://buildstream.gitlab.io/buildstream/projectconf.html#source-aliases) (see below for a description of an *alias mapping*
* There is no restriction on domains of *alias mappings* within a single *mirror*, although we *expect* that they are grouped together by geographical location, this also need not necessarily be true.
* A mirror definition is allowed to be *sparse*, such that not all source aliases defined by the project need to have an *alias mapping*
  * This is mostly a provision for projects which manage their own central VCS repositories for a hand full of sources
* A mirror may address and override *alias mappings* in *subprojects* accessed via *junctions*
  * This allows higher level projects to easily redirect fetches for subproject sources from a more efficient location

Structure of an *alias mapping*:
* An *alias mapping* is used to indicate one or more alternative URLs with which BuildStream can resolve a *source alias*
* The ability to map multiple aliases inside a single *mirror* definition is a provision to group multiple mirrors into a single logical *geographic location*, allowing fallback repositories for the cases where an upstream source has been modified, this is required for when:
  * An upstream tarball retains the same name, but was manually updated, such that the upstream URI remains the same, but the sha256sum has changed
  * An upstream VCS repository has undergone surgery, such that for example, a git commit sha which a project is using has disappeared from upstream by means of a history rewrite.
* There is an open question as to whether using multiple URIs in an *alias mapping* for the purpose described above should be a **rule** rather than just a **provision**, if we enforce this this structure as a rule, we can make some assumptions, such as:
  * The order of URIs in the *mapping* is meaningful, such that earlier URIs in the list represent older versions of the same repo, meaning that it is always appropriate to `bst track` with *later* URIs in the list

## Configuration example for `project.conf`

```yaml

aliases:
  foo-git: git://git.foo.com/git
  foo-tar: https://download.foo.com/sources


mirrors:
- location-name: united-kingdom
  aliases:
    foo-git:
    - git://git.foo.co.uk/git
    foo-tar:
    - https://download.foo.co.uk/sources0
    - https://download.foo.co.uk/sources1

    # Subproject "bar" gits
    bar:bar-git:
    - git://git.bar.co.uk/git

- location-name: korea
  aliases:
    foo-git:
    - git://git.foo.co.kr/git
    foo-tar:
    - https://download.foo.co.kr/sources0
    - https://download.foo.co.kr/sources1

    # Subproject "bar" gits
    bar:bar-git:
    - git://git.bar.co.uk/git
```

## Runtime behaviors

When running a `fetch` or `track` task for a given `Source`, multiple iterations must now be made such that we can try multiple aliases.

First, we can compose a normalized list of URIs to traverse for a given *source alias* somewhere in the pipeline initialization phase.

At `fetch` time, the general order should usually be:
* First try the default *mirror*
  * Iterate through *alias mappings* for the given *mirror*
  * Open question: While iterating through *alias mappings*, should we optimize by assuming that if we have not yet found the ref we are looking for in a `fetch` task, and that the repository is unreachable for the next URI, we can safely move on to the next *mirror* ? This would turn our expectation of how the *alias mappings* work into a rule.
* Try the remaining *mirrors* in the order they are declared in the `project.conf`
* Resort to the true upstream URL, i.e. default expansion of the *source alias* itself.

At `track` time, things are a bit more complicated, we may want to do the whole thing in reverse; such that we are guaranteed to always make an attempt to track the *latest* by default.

## Implementation of iterating over aliases

The `Source` plugin facing API should remain unchanged for this, existing plugins can continue to function without any added bells or whistles.

Initially, `Source` objects are instantiated in the main data model with *source aliases* resolved to the true upstream default URLs in the normal way.

During this initial instantiation for the main data model, we can infer *which source alias* was used by the given *Source* object by observing it's call to `Source.translate_url()`. This allows us to construct a list of URLs later on which should be *tried*.

Inside the `TrackQueue` and `FetchQueue` tasks, the Source may be *reinstantiated* multiple times, for each *try* in this context, we make `Source.translate_url()` resolve the URI to something different.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 29, 2018, 12:05

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 30, 2018, 19:39

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 30, 2018, 19:40

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 30, 2018, 19:47

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 30, 2018, 19:50

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 30, 2018, 19:52

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Apr 5, 2018, 18:13

Do I understand correctly that the way to make this work is to have translate_url do double-duty?

In the first case, when the pipeline is assembled and `configure()` is called on the sources, it stores somewhere (probably the Project) what aliases that source used.

In the second case, when the TrackQueue/FetchQueue is operating on an element, it substitutes the alias for an appropriate alias mapping.

Does buildstream have a clear way of determining which one this is? i.e.
* Is something in the project or pipeline already set when the pipeline has finished being initialised?
* Should we set something in the pipeline or project when we want the second case to happen?
* Is it enough to know that we've already stored what alias that source used, so that all subsequent calls are for the second case?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 6, 2018, 10:50

> Do I understand correctly that the way to make this work is to have translate_url do double-duty?

Yes.

> In the first case, when the pipeline is assembled and `configure()` is called on the sources, it stores somewhere (probably the Project) what aliases that source used.

I would think the `Source` itself is a better place to store this.

> In the second case, when the TrackQueue/FetchQueue is operating on an element, it substitutes the alias for an appropriate alias mapping.

> Does buildstream have a clear way of determining which one this is? i.e.

This is where BuildStream internally has to implement a loop, which may be different depending on fetch or on track task (lets start with fetch, the order of mirrors to check for fetch is more clearly defined, track is a bit more exotic).

> * Is something in the project or pipeline already set when the pipeline has finished being initialised?

> * Should we set something in the pipeline or project when we want the second case to happen?

> * Is it enough to know that we've already stored what alias that source used, so that all subsequent calls are for the second case?

When it comes time to try one of the mirrors and loop through potential URLs, we are in the main context in the `FetchQueue`, so we remember already which `alias` that source used (if the source neglected to use an alias, we just treat it in the regular way without any loop).

From here, we *instantiate a new `Source` object* for each iteration, under the hood we're telling that new `Source` instance which alternative mirror URL to resolve when `Source.translate_url()` gets called by the new instance.

We instantiate a new Source object for each iteration here.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Apr 6, 2018, 10:57

Ah, I get it! Since we're instantiating our own sources, we can pass into them when they're being constructed that we want to use an alias mapping.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Apr 6, 2018, 16:50

Looking at the way junctions and projects inter-operate, it seems that projects are not aware of their subprojects, though a project is indirectly aware of its superproject (project.junction.project).

Should the subprojects defined there be able to chain indefinitely (e.g. ``foo:baz:bar:bar-git``) or is one step enough?

Also, considering the proposal for include statements, would it be appropriate to use them instead of adding new functionality in this way?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Apr 9, 2018, 16:58

While trying to implement this, I've run into the issue that we don't seem to have enough information to instantiate a new source based on the old one - its constructor requires the `meta` arg, which is not stored in its entirety in the source.

I'll do further investigation as to where it comes from and whether it's easily recoverable, but it looks like the simplest solution is to have the source store `meta` as well.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Apr 16, 2018, 16:59

created branch [`328-support-for-downloading-sources-from-mirrors`](https://gitlab.com/BuildStream/buildstream/compare/master...328-support-for-downloading-sources-from-mirrors)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Apr 16, 2018, 17:11

mentioned in merge request !404

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on May 15, 2018, 14:05

marked this issue as related to #330

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on May 15, 2018, 14:12

assigned to [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on May 24, 2018, 13:57

Following discussion by E-mail, some conventions have been agreed on so that work on generating buildstream mirrors can happen in parallel:

* Sources can raise a RefNotFoundError(that inherits from SourceError) if the repository exists, but the desired ref was not found - this will be used to iterate through repositories in a generated mirror to find one that contains the ref.
* When fetching from buildstream-generated mirrors, I will use the format {URL}/{source_kind}/{URL_directory_name}/{mirror_number} to find the URL, where:
  - {URL} is the URL of the buildstream mirror generated via config,
  - {source_kind} is the kind of the source (as specified in the yaml's "kind" field),
  - {URL_directory_name} is the URL specified for the source, mangled into a valid name for a directory (e.g. git://github.com becomes git___github_com),
  - {mirror_number} as a number starting from 0, and incrementing every time fetching/tracking returns a RefNotFoundError instead of a general SourceError.
* There will be a helper method to make this format easier to use, probably of the form `Source.get_normalised_mirror_path(upstream_url, *, prefix="", suffix="")`

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on May 24, 2018, 14:05

However, on further investigation, I may have to make some changes.

1. {URL} includes a protocol type which may not work for fetching every source.
   i.e. http:// can be used to fetch from downloadable file sources (zip, tar), and git, but I doubt it will work for every source type that needs mirroring (i.e. do svn, bzr and hg sources accept http:// ?).

2. I previously expected Source.get_normalised_mirror_path(upstream_url, *, prefix="", suffix="") to use the "suffix" field to append the {mirror_number}. I don't think it will actually work that way, so I don't currently have a use for suffix.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on May 31, 2018, 13:35

This item seems to have gotten very complex, so I'm scaling back some of the complexity (though leaving the old WIP branch with the old changes intact for anyone who might use that functionality).

The branch https://gitlab.com/BuildStream/buildstream/commits/jonathan/mirror-client contains all of my work so far, including an under-tested attempt at fetching from incremental namespaces.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Jul 4, 2018, 18:13

Following discussions and review in !404 , the current state of this issue is that I am most of the way through an implementation of this that is backwards-compatible with old sources, though sources that fetch from multiple URLs will not use mirroring to its full extent (it might work when all the URLs use the same alias).

I still have to add proper documentation around this new approach, and add documentation on how to set up simple git and tar mirrors to make use of this functionality.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Jul 27, 2018, 14:08

closed via merge request !404

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jul 29, 2018, 09:10

mentioned in merge request !582

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Aug 3, 2018, 09:20

marked this issue as related to #537

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Aug 3, 2018, 09:20

marked this issue as related to #538
